### ISSUE TITLE
flycheck-eslint doesn't honour rules in ~/.eslintrc

### ISSUE THREAD

**Issue by sunwukung** [flycheck/flycheck#735]
I'm having trouble getting flycheck to use the .eslintrc file in the root of the current project, and I'm struggling to find any documentation that will tell me how to configure it to use the appropriate file. This setup was derived from [this blogpost](http://codewinds.com/blog/2015-04-02-emacs-flycheck-eslint-jsx.html)

    (use-package flycheck
      :ensure t
      :diminish flycheck-mode
      :config
      (global-flycheck-mode)
    
      (setq-default flycheck-temp-prefix ".")
      (setq-default flycheck-disabled-checkers
                    (append flycheck-disabled-checkers
                            '(javascript-jshint)))
      (setq flycheck-checkers '(javascript-eslint))
      (setq-default flycheck-disabled-checkers
                    (append flycheck-disabled-checkers
                            '(json-jsonlist)))
    (setq flycheck-eslintrc "~/.eslintrc"))  
    )
    
    (provide 'init-flycheck)

if I run eslint from the CLI it reports the errors I'd expect, but in emacs it just lists (what I assume are) the default errors regarding semicolon usage. Can anyone shed any light on this?

**Comment by cpitclaudel** [comment]
Hey sunwukung,

Could you try pressing <kbd>C-c ! C-c</kdb> in a javascript buffer? This will run eslint and should tell you what exact command line flycheck is using to call eslint. On my machine. after running `(setq flycheck-eslintrc "~/.eslintrc"))`. it shows something like 

```
eslint --format\=checkstyle --config /home/clement/.eslintrc /home/clement/test.js
```

By the way, the `(setq flycheck-checkers '(javascript-eslint))` line in your config disables flycheck for everything except for javascript. Do you actually want that?

**Comment by sunwukung** [comment]
Hi, 
Hmm - yup, this indeed worked as yours did - and the errors it listed were as I'd expect. However, flycheck doesn't indicate anything visually, nor does it list any of the errors if I hit C-c ! l 
I've tried flycheck-verify-checker, and it states that javascript-eslint can run in the current buffer, and I've tried flycheck-select-checker in case it wasn't getting automatically cued up by entering js2-mode.
You were right about that stray bit of config, I think that was lying about just as part of an experiment to try and force flycheck to use eslint. Here's an updated config..

    (use-package flycheck
      :ensure t
      :diminish flycheck-mode
      :config
      (global-flycheck-mode)
      (setq-default flycheck-temp-prefix ".")
      (setq flycheck-eslintrc "~/.eslintrc")
      (flycheck-add-mode 'javascript-eslint 'js2-mode)
      (flycheck-add-mode 'javascript-eslint 'js-mode)

      (setq-default flycheck-disabled-checkers
                    (append flycheck-disabled-checkers
                            '(javascript-jshint))))


    (provide 'init-flycheck)

**Comment by cpitclaudel** [comment]
Thanks for the extra info! It looks like Flycheck is actually finding your config then. Could you post the full output of <kbd>C-c ! v</kbd>? 

I don't think the orange underlines are a sign that flycheck is in fact finding errors: js-2 mode includes it's own little bit of error checking. What you are most likely seeing is js2's own errors.

Regarding your config, the following two lines shouldn't be needed, as `javascript-eslint` already supports `js-mode` and `js2-mode` out of the box.
```
  (flycheck-add-mode 'javascript-eslint 'js2-mode)
  (flycheck-add-mode 'javascript-eslint 'js-mode)
```

**Comment by sunwukung** [comment]
Here you go:

    Syntax checkers for buffer test.js in js2-mode:

      javascript-jshint
        - predicate:  nil
        - executable: Not found

      javascript-eslint
        - predicate:  t
        - executable: Found at /opt/boxen/nodenv/shims/eslint

      javascript-gjslint
        - predicate:  nil
        - executable: Not found

      javascript-jscs
        - predicate:  nil
       - executable: Not found

      javascript-standard
        - predicate:  nil
        - executable: Not found

Flycheck Mode is enabled.

**Comment by cpitclaudel** [comment]
Thanks! So it seems Flycheck is properly setup and enabled. At this point @lunaryorn will probably be able to help more than myself, but hopefully the information you gave will make it easier for him. It could be useful to add two things:

* Your .eslintrc
* A javascript file for which the output of <kbd>C-c ! C-c</kbd> contains errors or warnings, but flycheck shows nothing.

Then maybe I could reproduce that issue, and then I could probably help :)

Which version of emacs are you using?

**Comment by sunwukung** [comment]
Hi, thanks for your help - here's my .eslintrc 

    {
        "parser": "babel-eslint",
        "rules": {
            "strict": 0,
            "indent": [
                2,
                2
            ],
            "quotes": [
                2,
                "double"
            ],
            "linebreak-style": [
                2,
                "unix"
            ],
            "semi": [
                2,
                "always"
            ]
        },
        "env": {
            "es6": true,
            "browser": true
        }
    }

it's pretty simplistic - I pared it right back to try and eliminate any issues that might originate from the config. Here's a sample test file I've been using: 

    function foo() {
      var api = {
        value: function() {
          var x = 'badstring'; // double quotes rules should kick in here
        }
      };
    }

    export {foo}

Here's the output from running C-c ! C-c

    -*- mode: compilation; default-directory: "~/Documents/code/" -*-
    Compilation started at Wed Aug 26 09:45:44

    eslint --format\=checkstyle --config /Users/sunwukung/.eslintrc /Users/sunwukung/Documents/code/test.js
    <?xml version="1.0" encoding="utf-8"?><checkstyle version="4.3"><file name="/Users/sunwukung/Documents/code/test.js"><error line="4" column="15" severity="error" message="Strings must use doublequote. (quotes)" /><error line="9" column="13" severity="error" message="Missing semicolon. (semi)" /></file></checkstyle>

    Compilation exited abnormally with code 1 at Wed Aug 26 09:45:50

**Comment by sunwukung** [comment]
OK, this might be something of a red herring - I just initialised flycheck using this setup from someone over on reddit - and flycheck starts working as expected. Given my original configuration, could you shed some light on why this works (I'm a complete noob at elisp)

    (add-hook 'js2-mode-hook
          (defun my-js2-mode-setup ()
            (flycheck-mode t)
            (when (executable-find "eslint")
              (flycheck-select-checker 'javascript-eslint))))

**Comment by sunwukung** [comment]
bingo! it's this line:
    
    (setq-default flycheck-temp-prefix ".")

this was added to my config following a previous discussion with lunaryorn to try and hide files from processes like mocha that watch files in folders. Is there something wrong with the way that I have attempted to set the prefix? Or is it perhaps preventing flycheck itself from seeing it's reports?

**Comment by cpitclaudel** [comment]
For the record, I think the problem is actually with eslint. Indeed, for this file saved as `test.js`:

```eslint
var foo = bar;
```

and this `.eslintrc`:

```
extends:
    "eslint:recommended"
```

I get the following errors:

```
<?xml version="1.0" encoding="utf-8"?><checkstyle version="4.3"><file name="/home/clement/test.js"><error line="1" column="5" severity="error" message="foo is defined but never used (no-unused-vars)" source="eslint.rules.no-unused-vars" /><error line="1" column="11" severity="error" message="&quot;bar&quot; is not defined. (no-undef)" source="eslint.rules.no-undef" /></file></checkstyle>
```

However, when I name the input `.test.js`, I get:

```
<?xml version="1.0" encoding="utf-8"?><checkstyle version="4.3"><file name="/home/clement/.test.js"><error line="undefined" column="undefined" severity="warning" message="File ignored because of your .eslintignore file. Use --no-ignore to override." source="" /></file></checkstyle>
```

When flycheck uses "." as its prefix, it gets the same response from `eslint`. Since that message has no line number, `flycheck-relevant-error-p` filters it out. 

@lunaryorn, should we keep that error and display it in the error list, even if it doesn't have a line? It would require changes in a number of places, as for example `-add-overlay` does expect errors to have a location.

Or perhaps eslint should just return 1, 1 for the position of that error?

**Comment by lunaryorn** [comment]
@cpitclaudel What happens if you add `--no-ignore`?

**Comment by cpitclaudel** [comment]
Then it returns no errors at all. I wonder if I'm doing something wrong:

```sh
$ eslint --format\=checkstyle .test.js
<?xml version="1.0" encoding="utf-8"?><checkstyle version="4.3"><file name=".test.js"><error line="undefined" column="undefined" severity="warning" message="File ignored because of your .eslintignore file. Use --no-ignore to override." source="" /></file></checkstyle>
$ cat .eslintignore
cat: .eslintignore: No such file or directory
01:31:29 /home/clement
$ eslint --format\=checkstyle --no-ignore .test.js
<?xml version="1.0" encoding="utf-8"?><checkstyle version="4.3"></checkstyle>
$ cp .test.js test.js
$ eslint --format\=checkstyle test.js
<?xml version="1.0" encoding="utf-8"?><checkstyle version="4.3"><file name="test.js"><error line="1" column="5" severity="error" message="foo is defined but never used (no-unused-vars)" source="eslint.rules.no-unused-vars" /><error line="1" column="11" severity="error" message="&quot;bar&quot; is not defined. (no-undef)" source="eslint.rules.no-undef" /></file></checkstyle>
$ eslint --format\=checkstyle --no-ignore test.js
<?xml version="1.0" encoding="utf-8"?><checkstyle version="4.3"><file name="test.js"><error line="1" column="5" severity="error" message="foo is defined but never used (no-unused-vars)" source="eslint.rules.no-unused-vars" /><error line="1" column="11" severity="error" message="&quot;bar&quot; is not defined. (no-undef)" source="eslint.rules.no-undef" /></file></checkstyle>
```

**Comment by cpitclaudel** [comment]
Never mind, that's indeed an upstream bug at https://github.com/eslint/eslint/blob/3de24f103301138171094333d89be51f30839c87/lib/util/traverse.js#L44

**Comment by cpitclaudel** [comment]
The place where that message is printed is https://github.com/eslint/eslint/blob/3de24f103301138171094333d89be51f30839c87/lib/cli-engine.js#L398

**Comment by lunaryorn** [comment]
@cpitclaudel Seems to be an upstream issue then. Apparently eslint just refuses to check hidden files, for whatever reason.  I don't think we should go to great length to work around it, and rather just report it to eslint.  Specifically I think that eslint should not refuse to check hidden files if they are explicitly given on command line.

**Comment by cpitclaudel** [comment]
Sure, I'll open an upstream issue.

**Comment by lunaryorn** [comment]
@cpitclaudel Great, thank you!

**Comment by cpitclaudel** [comment]
Done!
